### DISCUSSION TITLE
Why are "ON" and "OFF" keys not configurable as a input key in keymap editor?

### DISCUSSION THREAD

**Discussion by soya-daizu** [google/mozc#627] [Q&A] [answered]
Environment: Manjaro Linux, Fcitx5, Mozc-2.28.4830.102+24.11.oss

I wanted to set IME ON and OFF keys (internally recognized as Hangul and Hangul_Hanja respectively) for switching between Kana and Eisu input mode to reproduce the behavior in macOS, but it appears like these keys are not allowed to be used as a input key.
What is the preffered way to do this in Mozc? While Fcitx can entirely switch off from Mozc with key bindings, I want to use the direct input mode of Mozc instead of disabling Mozc.

![image](https://user-images.githubusercontent.com/18283756/199018496-898534af-f388-46be-af32-cf59c557837b.png)

The error dialog appears at:
https://github.com/google/mozc/blob/master/src/gui/config_dialog/keymap_editor.cc#L443-L448

ON and OFF keys are included in `invisible_key_events_` at:
https://github.com/google/mozc/blob/master/src/gui/config_dialog/keymap_editor.cc#L124-L125

Thus not accepting the keys in:
https://github.com/google/mozc/blob/master/src/gui/config_dialog/keymap_editor.cc#L143-L148

**Comment by hiroyuki-komatsu** [comment, marked as answer]
Hi soya-daizu,

IME ON and OFF are predefined keys to turn on and off the IME.
So the users don't need to explicitly set up them, and the users cannot modify them from turn on and off the IME.

Since https://github.com/google/mozc/issues/552 (https://github.com/google/mozc/commit/80bc73f06dcd4e1a52a03f46c2ae308f3e1bddb6), Kana (Hangul) / Eisu (Hangul_Hanja) are treated as ON and OFF on Linux.

If ibus-mozc works on your environment, you might need to configure Fcitx5 to send ON and OFF on pressing Kana and Eisu.
Thanks,
